runtime/jcl/common/java_lang_invoke_MethodHandleNatives.cpp

```diff
@@ -0,0 +1,1434 @@
+			flags |= MH_REF_INVOKESPECIAL << MN_REFERENCE_KIND_SHIFT;
+		} else {
+			flags |= MH_REF_INVOKEVIRTUAL << MN_REFERENCE_KIND_SHIFT;
+		}
+
+		clazzObject = J9VMJAVALANGREFLECTMETHOD_DECLARINGCLASS(currentThread, refObject);
+	} else if (refClass == J9VMJAVALANGREFLECTCONSTRUCTOR(vm)) {
+		J9JNIMethodID *methodID = vm->reflectFunctions.idFromConstructorObject(currentThread, refObject);
+		vmindex = (jlong)methodID;
+		target = (jlong)methodID->method;
+
+		J9ROMMethod *romMethod = J9_ROM_METHOD_FROM_RAM_METHOD(methodID->method);
+
+		flags = romMethod->modifiers & CFR_METHOD_ACCESS_MASK;
+		flags |= MN_IS_CONSTRUCTOR | (MH_REF_INVOKESPECIAL << MN_REFERENCE_KIND_SHIFT);
+
+		clazzObject = J9VMJAVALANGREFLECTMETHOD_DECLARINGCLASS(currentThread, refObject);
+	} else {
+		vmFuncs->setCurrentExceptionUTF(currentThread, J9VMCONSTANTPOOL_JAVALANGILLEGALARGUMENTEXCEPTION, NULL);
+	}
+
+	if (!VM_VMHelpers::exceptionPending(currentThread)) {
+		J9VMJAVALANGINVOKEMEMBERNAME_SET_FLAGS(currentThread, membernameObject, flags);
+		J9VMJAVALANGINVOKEMEMBERNAME_SET_CLAZZ(currentThread, membernameObject, clazzObject);
+		J9OBJECT_U64_STORE(currentThread, membernameObject, vm->vmindexOffset, (U_64)vmindex);
+		J9OBJECT_U64_STORE(currentThread, membernameObject, vm->vmtargetOffset, (U_64)target);
+		Trc_JCL_java_lang_invoke_MethodHandleNatives_initImpl_setData(currentThread, flags, clazzObject, vmindex, target);
+	}
+}
+
+static char *
+sigForPrimitiveOrVoid(J9JavaVM *vm, J9Class *clazz)
+{
+	char result = 0;
+	if (clazz == vm->booleanReflectClass)
+		result = 'Z';
+	else if (clazz == vm->byteReflectClass)
+		result = 'B';
+	else if (clazz == vm->charReflectClass)
+		result = 'C';
+	else if (clazz == vm->shortReflectClass)
+		result = 'S';
+	else if (clazz == vm->intReflectClass)
+		result = 'I';
+	else if (clazz == vm->longReflectClass)
+		result = 'J';
+	else if (clazz == vm->floatReflectClass)
+		result = 'F';
+	else if (clazz == vm->doubleReflectClass)
+		result = 'D';
+	else if (clazz == vm->voidReflectClass)
+		result = 'V';
+
+	if (result) {
+		PORT_ACCESS_FROM_JAVAVM(vm);
+		char* signature = (char*)j9mem_allocate_memory(2 * sizeof(char*), OMRMEM_CATEGORY_VM);
+		signature[0] = result;
+		signature[1] = '\0';
+		return signature;
+	}
+	return NULL;
+}
+
+char *
+getClassSignature(J9VMThread *currentThread, J9Class * clazz)
+{
+	PORT_ACCESS_FROM_JAVAVM(currentThread->javaVM);
+
+	U_32 numDims = 0;
+
+	J9Class * myClass = clazz;
+	while (J9ROMCLASS_IS_ARRAY(myClass->romClass))
+	{
+		J9Class * componentClass = (J9Class *)(((J9ArrayClass*)myClass)->componentType);
+		if (J9ROMCLASS_IS_PRIMITIVE_TYPE(componentClass->romClass)) {
+			break;
+		}
+		numDims++;
+		myClass = componentClass;
+	}
+	J9UTF8 * romName = J9ROMCLASS_CLASSNAME(myClass->romClass);
+	U_32 len = J9UTF8_LENGTH(romName);
+	char * name =  (char *)J9UTF8_DATA(romName);
+	U_32 length = len + numDims;
+	if (* name != '[')
+		length += 2;
```

Comment:
There's a lot of missing `{` & `}`.  Please always use them for control flow as it prevents errors when modifying the code.
```suggestion
	if (* name != '[') {
		length += 2;
	}
```